### PR TITLE
Adopt Sender.append() and flush()

### PR DESCRIPTION
**Pull request by rmfitzpatrick**
Implements the `Sender.append()` and `Sender.flush()` methods requested in https://github.com/jaegertracing/jaeger-client-python/pull/186#pullrequestreview-125400053 in an attempt to get those changes sourced.

Signed-off-by: Ryan Fitzpatrick <rmfitzpatrick@signalfx.com>

### CODE REVIEW

**Comment by rmfitzpatrick** [comment]
@ProvoK, if you're interested in continuing with these changes, do you mind rebasing your branch to update the dependencies for successful CI runs?

**Comment by ProvoK** [comment]
@rmfitzpatrick Hello there,

thanks for continuing the work! Yes I'm interested, I was a lot busy in the last period, that's why I was not working on it.

I'll update my branch and review your change asap 😉

**Comment by ProvoK** [comment]
@rmfitzpatrick I've fetched upstream master and merged into my branch. This is all I can do till Monday, sorry!

File: jaeger_client/reporter.py
Comment (ProvoK) [inline jaeger_client/reporter.py]: Keeping the count of an internal state of the sender outside of it doesn't seem to me like a good idea.
It's better that the sender expose a property like `spans_count` which returns the real and actual count of its spans.
With that, you can easily get rid of updating this `spans_appended` variable (line 183, 187), with a more secure and less error prone behaviour.

File: jaeger_client/senders.py
Comment (ProvoK) [inline jaeger_client/senders.py]: I'm not very skilled with tornado: is there a valid reason to handle exceptions in this manner?

File: jaeger_client/senders.py
Comment (ProvoK) [inline jaeger_client/senders.py]: This docstring does not belongs to this method

File: jaeger_client/senders.py
Comment (ProvoK) [inline jaeger_client/senders.py:36]: `host` and `port` are likely to be needed in every sender implementations. I would keep them in the base class

**Review by ProvoK — CHANGES_REQUESTED** [review]

File: jaeger_client/senders.py
Comment (rmfitzpatrick) [reply jaeger_client/senders.py]: This is a special exception that allows coroutines to return values in older versions of python: http://www.tornadoweb.org/en/stable/gen.html#tornado.gen.Return.  It's necessary to get the exception info from the send generator.

File: jaeger_client/senders.py
Comment (rmfitzpatrick) [reply jaeger_client/senders.py:36]: In my early HTTPSender implementation I found only a url is necessary, and constructing one via host, port, and trace endpoint seems overkill (other jaeger clients use `JAEGER_ENDPOINT` env var to specify the collector location).

File: jaeger_client/reporter.py
Comment (rmfitzpatrick) [reply jaeger_client/reporter.py]: Great idea, updated.

File: jaeger_client/senders.py
Comment (rmfitzpatrick) [reply jaeger_client/senders.py]: Good call. Updated.

**Comment by rmfitzpatrick** [comment]
@ProvoK, thanks for the feedback, which I believe I've addressed as requested with the exception of https://github.com/ProvoK/jaeger-client-python/pull/1#discussion_r202802357 to match the logic of other clients.  Let me know if you still request that change and I'll be happy to do so, or if you have any other notes.

File: jaeger_client/senders.py
Comment (ProvoK) [reply jaeger_client/senders.py]: Ok, I got it. Still I'm not sure this is necessary. 

The only return value used is `num_spans`, which is everytime equal to `sender.span_count` with the actual implementation. 

This means that you can simply save a variable [here](https://github.com/ProvoK/jaeger-client-python/pull/1/files#diff-8a53213519652bcef509b8fbddbdfe86R184) (something like `num_spans = self._sender.span_count`) and then use it later for reporting the number of success/error spans.

Everything else (exceptions and theirs messages) can be handled normally, I suppose.
@rmfitzpatrick

File: jaeger_client/senders.py
Comment (ProvoK) [reply jaeger_client/senders.py:36]: Ok, let's see if jaeger maintainers are ok with this interface

Comment (rmfitzpatrick) [reply jaeger_client/senders.py:36]: Do you mean before merging into your PR branch?

Comment (ProvoK) [reply jaeger_client/senders.py:36]: Nono, I mean after merging in my PR branch. The only one remaining issue is the one related to `tornado.gen.Return`, I replied to you there yesterday

Comment (rmfitzpatrick) [reply jaeger_client/senders.py:36]: Sorry, I missed that comment.

File: jaeger_client/senders.py
Comment (rmfitzpatrick) [reply jaeger_client/senders.py]: Of course, makes more sense to fully adopt span_count and I've updated.  However, I did retain the `Sender._flush()` generated exception info so that transport-specific exception handling is possible in Sender and its subclasses, without having to maintain a complete, sender-aware try/except block in `Reporter._consume_queue()`.  Does that work for you, @ProvoK?

File: jaeger_client/reporter.py
Comment (ProvoK) [inline jaeger_client/reporter.py]: I don't get the benefit of that, honestly. What you are doing now is a `try ...except Exception ...` in the `Sender.flush()`, returning the exception plus a message.  Which is the same as an `raise AnyException("message")`

In the `Reporter` you just check if any kind of `exc` is raised, and the use the message returned. 

This is the very same of:
```python
try:
    yield self._sender.flush()
except Exception as exc:
   self.metrics.reporter_failure(num_spans)
   self.error_reporter.error(error_msg, exc)
else:
    self.metrics.reporter_success(num_spans)
```
but this is the most common and standard way to deal with exceptions, in Python.

The point is to keep the code "idiomatic".
In other languages, as Go, the standard is to have the pattern you are doing: multiple returns, one for the error/ok-boolean and the other for the "real" value.

File: jaeger_client/senders.py
Comment (ProvoK) [inline jaeger_client/senders.py]: Here you can just `raise e("my message") from e` and it's done! (alongwith the previous comment modifications)

The `from e` part preserve the exception stacktrace

**Review by ProvoK — CHANGES_REQUESTED** [review]

File: jaeger_client/senders.py
Comment (ProvoK) [inline jaeger_client/senders.py]: Again here, you can save a lot of lines of code

**Review by ProvoK — CHANGES_REQUESTED** [review]

File: jaeger_client/senders.py
Comment (rmfitzpatrick) [reply jaeger_client/senders.py]: `from e` is not supported in python 2.

File: jaeger_client/reporter.py
Comment (rmfitzpatrick) [reply jaeger_client/reporter.py]: Where does `error_msg` come from in this excerpt?  Providing `error_reporter.error()` with `str(exc)` and `exc` loses the information provided by the initial local-agent submission failure: https://github.com/jaegertracing/jaeger-client-python/blob/master/jaeger_client/reporter.py#L201

This pattern could be repeated with all manner of clients in the future, who idiomatically catch their transport specific exceptions and provide a relevant exception message.

Comment (ProvoK) [reply jaeger_client/reporter.py]: I was writing in this little github editor so I lost the context
```python
try:
    yield self._sender.flush()
except Exception as exc:
   self.metrics.reporter_failure(num_spans)
   self.error_reporter.error("Failed to submit traces to jaeger: {}".format(exc), exc)
else:
    self.metrics.reporter_success(num_spans)
```

You can still catch below, in the sender, the specific error and provide a relevant message, carried with the exception itself.

File: jaeger_client/senders.py
Comment (ProvoK) [reply jaeger_client/senders.py]: Yeah, you're right. I'm too used to python3

File: jaeger_client/reporter.py
Comment (rmfitzpatrick) [reply jaeger_client/reporter.py]: Providing a custom message in `send()` seems effective.  Are you comfortable with using something like `six.reraise` to add a custom value with the traceback from `sys.exc_info()`?

Comment (ProvoK) [reply jaeger_client/reporter.py]: I would not add six as dependency just for that. And probably jaeger maitaners wouldn't too

Comment (rmfitzpatrick) [reply jaeger_client/reporter.py]: It already is one via thrift.

edit:  Not to speak in circles, but I'm not aware of a version agnostic way to override an exception value without making Reporter sender detail aware or less informative otherwise.

Comment (ProvoK) [reply jaeger_client/reporter.py]: Oh, didn't know it was included in thrift. Then I suppose it's allright to use it

**Comment by rmfitzpatrick** [comment]
edit:  User error with a stale test.

I really appreciate your thorough feedback and believe I've addressed it all, @ProvoK.

**Review by ProvoK — APPROVED** [review]

**Comment by ProvoK** [comment]
Thanks @rmfitzpatrick for your work, I'm going to merge this, and I'll quote in in the original PR.

**Pull request merged**
